README
CloudFlare Keyless SSL reference implementation user guide
==========================================================

This repository contains a reference implementation of CloudFlare's
keyless SSL server.

Protocol
--------

The CloudFlare Keyless SSL client communicates to the server via a
binary protocol over a mutually authenticated TLS tunnel.  Messages
are in binary format and identified by a unique ID.

Messages consist of a fixed length header, and a variable length body.
The body of the message consists of a sequence of items in TLV (tag,
length, value) messages.

All messages with major version 1 will conform to the following
format.  The minor version is currently set to 0 and is reserved for
communicating policy information.

Header:

    0 - - 1 - - 2 - - 3 - - 4 - - - - 6 - - 7 - - 8
    | Maj | Min |   Length  |          ID           |
    |                    Body                       |
    |     Body     | <- 8 + Length

Item:

    0 - - 1 - - 2 - - 3 - - 4 - - - - 6 - - 7 - - 8
    | Tag |   Length  |          Data               |
    |           Data             | <- 3 + Length

All numbers are in network byte order (big endian).

The following tag values are possible for items:

    0x01 - Certificate Digest,
    0x02 - Server Name Indication,
    0x03 - Client's IP address,
    0x11 - Opcode,
    0x12 - Payload,

A requests contains a header and the following items:

    0x01 - length: 32 bytes, data: SHA256 of RSA modulus
    0x02 - length: variable, data: SNI string
    0x03 - length: 4 or 16 bytes, data: IPv4/6 address
    0x11 - length: 1, data: opcode describing operation
    0x12 - length: variable, data: payload to sign or encrypt

The following opcodes are supported in the opcode item:

    0x01 - operation: RSA decrypt payload 
    0x02 - operation: RSA sign MD5SHA1
    0x03 - operation: RSA sign SHA1
    0x04 - operation: RSA sign SHA224
    0x05 - operation: RSA sign SHA256
    0x06 - operation: RSA sign SHA384
    0x07 - operation: RSA sign SHA512

Responses contain a header with a matching ID and only two items:

    0x11 - length: 1, data: opcode describing operation status
    0x12 - length: variable, data: payload response

The following opcodes are supported in the opcode item:

    0xF0 - operation: success, payload: modified payload
    0xFF - operation: RSA decrypt payload, payload: 

On an error, these are the possible 1-byte payloads:

    0x01 - cryptography failure
    0x02 - key not found - no matching certificate ID
    0x03 - read error - disk read failure
    0x04 - version mismatch - unsupported version incorrect
    0x05 - bad opcode - use of unknown opcode in request
    0x06 - unexpected opcode - use of response opcode in request
    0x07 - format error - malformed message
    0x08 - internal error - memory or other internal error

Key Management
--------------

The Keyless SSL server is a TLS server and therefore requires
cryptographic keys. All requests are mutually authenticated, so both
the client and the server need a TLS 1.2 compatible key pair.

The server will need a valid key and certificate pair in PEM format.
The following options are required and take a path to these files:

     --server-cert 
     --server-key

The private keys that this server is able to use should be stored in
PEM format in a directory denoted by the option:

    --private-key-directory

In order to authenticate the client'certificate, a custom CA file is
required.  This CA file is provided by CloudFlare and provided with:

    --ca-file

Code Organization
-----------------

The code is split into several files by function in order to enable
swapping with custom implementations.

    kssl.h              contains the shared constants and structures
    kssl_core.h         APIs for performing the keyless operation
    kssl_helpers.h      APIs for serialization and parsing functions
    kssl_private_key.h  APIs for storing and matching private keys
    kssl_log.h          APIs for writing logs

    kssl_server.c       sample server implementation with openssl and libuv
    kssl_testclient.c   client implementation with openssl

The following files are reference implementations of the APIs above.

    kssl_core.c         implementation of v1.0 policy for keyless operation
    kssl_helpers.c      implementation of v1 serialization and parsing
    kssl_private_key.c  implementation of reading, storage and operations of
                        private keys using openssl
    kssl_log.c          implementation of logging to stderr

Building and Dependencies
-------------------------

The Keyless SSL server implementation has two external dependencies,
OpenSSL and libuv.  These are open source and available for most
platforms.

For Unix-based systems, the server and test suite are built with a
Makefile.

To build:

    make

Result:

    o/kssl_testclient o/kssl_server

To test:

    make test

Recommended Settings
--------------------

The kssl_server command line options determine configuration.

The recommended value for cipher-list is: ECDHE-RSA-AES128-SHA256
The recommended port is: 24008

Keyserver Configuration
-----------------------

This is the keyserver for Keyless SSL. It consists of a single binary
file 'kssl_server' that has the following command-line options:

    --port
              (optional) The TCP port on which to listen for connections.
              These connections must be TLSv1.2. Defaults to 2407.

     --ip     
              (optional) The IP address of the interface to bind to.
              If missing binds to all available interfaces.
    
     --ca-file

              Path to a PEM-encoded file containing the CA certificate
              used to sign client certificates presented on connection.

    --server-cert
    --server-key

              Path to PEM-encoded files containing the certificate and
              private key that are used when a connection is made to the
              server. These must be signed by an authority that the client
              side recognizes (e.g. the same CA as --ca-file).

    --cipher-list

              An OpenSSL list of ciphers that the TLS server will accept
              for connections. e.g. ECDHE-RSA-AES128-SHA256:RC4:HIGH:!MD5

    --private-key-directory

              Path to a directory containing private keys which the keyserver
              provides decoding service against. The key files must end with
              ".key" and be PEM-encoded. There should be no trailing / on 
              the path.

    --silent
              Prevents keyserver from producing any log output. Fatal
              start up errors are sent to stderr.

    --verbose 

              Enables verbose logging. When enabled access log data is
              sent to the logger as well as errors.

    --num-workers

              (optional) The number of worker threads to start. Each worker
              thread will handle a single connection from a KSSL client. 
              Defaults to 1.

    --pid-file

              (optional) Path to a file into which the PID of the keyserver.
              This file is only written if the keyserver starts successfully.


The following options are not available on Windows systems:

    --user

            (optional) user:group to switch to. Can be in the form user:group
            or just user (in which case user:user is implied) (root only)

    --daemon

            (optional) Forks and abandons the parent process.

    --syslog

            (optional) Log lines are sent to syslog (instead of stdout
            or stderr). 

For example,

    kssl_server --port=2407                        \
                --server-cert=server-cert/cert.pem \
                --server-key=server-cert/key.pem   \
                --private-key-directory=keys       \
                --cipher-list=ECDHE-RSA-AES128-SHA256:AES128-GCM-SHA256:RC4:HIGH:!MD5:!aNULL:!EDH \
                --ca-file=CA/cacert.pem            \
                --pid-file=keyserver.pid

Makefile Targets
----------------

The Makefile has the following useful targets:

    all     The default target that builds both the keyserver and
            the testclient

    clean   Deletes the keyserver, testclient and related object files

    run     Runs the keyserver with a configuration suitable for testing
            (with the testclient)

    kill    Stops the keyserver started by 'make run'

    test    Runs the testclient against the keyserver.


RELEASE_NOTES
2015.2.0
- 2014-12-09 KEY-56 - Checkout do_ssl_handshake error and close if fatal
- 2014-12-15 Handle SSL_shutdown returning WANT_READ / WANT_WRITE
- 2014-12-15 Merge pull request #93 from jgrahamc/master
- 2014-12-16 Remove useless logging statement
- 2014-12-16 Merge pull request #95 from jgrahamc/master
- 2014-12-16 Re-ordered usage information such that required options are called out and at the top.
- 2014-12-16 Merge pull request #96 from buro9/master
- 2015-01-13 updated openssl_version to 1.0.1k
- 2015-01-13 Merge pull request #97 from cloudflare/update-openssl
- 2015-02-18 Update to use the one 5-year CA
- 2015-02-19 Merge pull request #98 from cloudflare/nick/5yearca
- 2015-02-19 updated makefile to use cfsetupv2 release tags

R1.6.1-20141125-1226
- 2014-11-12 Update the CA cert for compatibility.
- 2014-11-18 Merge pull request #89 from cloudflare/nick/CA-compat
- 2014-11-18 Added special keys to be installed for testing.
- 2014-11-19 Merge pull request #90 from cloudflare/jacob/testing-keys
- 2014-11-24 Bump OpenSSL to 1.0.1j
- 2014-11-25 Merge pull request #92 from cloudflare/nick/openssl


R1.6.0-20141016-0738
- 2014-06-26 added dockerfiles for centOS and debian
- 2014-06-26 added rpm-build to centos image
- 2014-06-29 Update readme with new build instructions
- 2014-06-29 KEY-23 - Tighten security on /etc/keyless/keys
- 2014-07-02 replaced openssl url with mirror, debian docker container has issues downloading from openssl.org
- 2014-09-10 Add an explanatory comment about why we loop around inside do_ssl
- 2014-09-09 Updated authoritative cipher_list and added ECDSA curve (NIST P-256). Removed --cipher-list command line argument.
- 2014-09-10 New Keyless and Keyserver CAs and consolidation of testing directory structure
- 2014-09-10 Updated tests to run againt servers with ECDSA and RSA certificates (with ECDSA clients)
- 2014-09-10 Print errno when the TLS handshake fails in the testclient
- 2014-09-11 Fixes for Windows compilation including --version segfault
- 2014-09-12 Warning not to use testing CAs in production and explanation of server certificate signing
- 2014-09-19 Merge pull request #72 from cloudflare/cipher-suites
- 2014-09-19 Merge pull request #68 from cloudflare/openssl-mirror
- 2014-09-19 Merge pull request #67 from cloudflare/dane
- 2014-09-19 Merge pull request #66 from cloudflare/dockerfiles
- 2014-09-19 Added IPv6 code into free-port.pl
- 2014-09-19 Attempt at making the code style more consistent.
- 2014-09-19 Merge pull request #74 from nitrix/fixMinorStyling
- 2014-09-19 Added documentation to code about -6 flag
- 2014-09-19 Update included openssl version
- 2014-09-20 Merge pull request #73 from mahtin/patch-1
- 2014-09-20 Merged windows branch and master
- 2014-09-21 Merge pull request #75 from cloudflare/windows
- 2014-09-22 Address issue #76 - check error return codes
- 2014-09-22 Address issue #77 - Remove unnecessary null pointer checks on free()
- 2014-09-22 Address issue #81 - Unsafe threaded use of ctime()
- 2014-09-22 Address issue #81 - Data race around gethostbyname in test code
- 2014-09-22 Address issue #76 - More mallocs() that were assumed to work
- 2014-09-22 Address issue #76 - more return parameter checking
- 2014-09-22 Address issue #76 - Error in checking error on fclose
- 2014-09-23 Address issue #80 - Add note about license not being 'open source'
- 2014-09-25 bugfix: disabled dtrace in the libuv build because it caused issues when recent versions of systemtap is installed.
- 2014-09-26 Merge pull request #84 from agentzh/no-dtrace
- 2014-09-26 Make -6 completely optional by using Perl require instead of use.
- 2014-09-26 Remove dependence on free-port.pl
- 2014-09-16 Generalized kssl_private_key to use EVP_keys, added support for EC keys.
- 2014-09-18 Cleanup testclient RSA tests
- 2014-10-10 Tests for ECDSA signing
- 2014-10-10 Update opcode documentation
- 2014-10-12 Merge pull request #87 from cloudflare/ec-keys
- 2014-10-13 Update bundled Keyless CA certificate in packages
- 2014-10-13 Key-44 Raw RSA Decrypt operation
- 2014-10-16 Merge pull request #88 from cloudflare/raw-rsa


R1.5.0-20140909-1415
- 2014-06-10 Update keyless.default
- 2014-06-12 Merge branch 'master' of github.com:cloudflare/keyless
- 2014-06-12 Pad good/bad response to 4,096 bytes
- 2014-06-13 Change from 4,096 to 1,024 byte padding since that's plenty of room
- 2014-06-25 KEY-19 - Reload keys on SIGHUP
- 2014-07-17 Add -dev to the version if this is not a release
- 2014-08-09 Fixed bug of processing pipelined keyless requests.
- 2014-08-09 Handle SSL internal error from RSA decryption.
- 2014-08-15 add SIGPIPE handler
- 2014-08-16 add pipeline request test
- 2014-08-19 Merge pull request #65 from cloudflare/KEY-19
- 2014-09-08 Merge pull request #69 from cdnsec/bugfix
- 2014-09-09 Fix broken pipelining when more than 16 messages queued


R1.4.0-20140609-1017
- 2014-06-09 Fix bad version number because tags not pulled


R1.3.0-20140609-1014
- 2014-05-14 Ensure that 'make kill' always succeeds and cleanup after test suite
- 2014-05-14 Merge pull request #52 from jgrahamc/master
- 2014-05-14 keyless release R1.3.0-20140514-0940
- 2014-05-14 keyless release R1.4.0-20140514-0956
- 2014-05-15 Remove the need to use fakeroot in packaging.
- 2014-05-15 Update read me to make markdown and add more details on on how to build.
- 2014-05-16 Improve documentation
- 2014-05-17 KEY-16 - Adding RPM packaging
- 2014-05-19 Add LICENSE file
- 2014-05-20 Use distro-specific makefiles for packaging
- 2014-05-20 Update package to mention license file
- 2014-05-21 KEY-16 fix bug in RPM packaging
- 2014-05-21 KEY-16 fix bug in packaging and add cf-package for internal build
- 2014-05-21 KEY-16 fix bugs in packaging
- 2014-05-21 KEY-16 config files in packages are relative
- 2014-05-22 KEY-16 more config file clean up
- 2014-05-22 Fix missing comments in defaults file
- 2014-05-22 KEY-16 - fix init script to just rely on keyless server to downgrade permissions

kssl.h
// kssl.h: definitions for the Keyless SSL protocol
//
// Copyright (c) 2013-2014 CloudFlare, Inc.

#ifndef INCLUDED_KSSL
#define INCLUDED_KSSL 1

#include <stdint.h>

#define BYTE uint8_t
#define WORD uint16_t
#define DWORD uint32_t

// These macros define the size of the kssl_header and kssl_item
// structures when they are serialized onto the wire (see
// kssl_helpers.c for seralization functions). These are not the same
// thing as sizeof(kssl_header) and sizeof(kssl_item) because
// of structure packing; the structure are used on deserialized
// data and are not direct representations of the wire protocol.
#define KSSL_HEADER_SIZE (sizeof(BYTE) + sizeof(BYTE) + sizeof(WORD) + sizeof(DWORD))
#define KSSL_ITEM_HEADER_SIZE (sizeof(BYTE) + sizeof(WORD))

typedef struct {
  BYTE version_maj; // Protocol major version (see KSSL_VERSION_MAJ)
  BYTE version_min; // Protocol minor version (see KSSL_VERSION_MIN)
  WORD length;      // Length of the payload
  DWORD id;         // Unique ID generated by client for this operation
  BYTE *data;       // Message contents TLVs
} kssl_header;

// The current KSSL protocol version
#define KSSL_VERSION_MAJ 0x01
#define KSSL_VERSION_MIN 0x00

// Possible item tags

#define KSSL_TAG_DIGEST     0x01 // An public key digest (see
                                 // digest_public_key)
#define KSSL_TAG_SNI        0x02 // Server name (optional)
#define KSSL_TAG_CLIENT_IP  0x03 // Client IP (4 bytes for IPv4, 16 for IPv6)
#define KSSL_TAG_OPCODE     0x11 // Requested operation (one of KSSL_OP_*)
#define KSSL_TAG_PAYLOAD    0x12 // Payload

#define KSSL_TAG_PADDING    0x20 // Padding

// Number of bytes to pad responses to

#define KSSL_PAD_TO 1024

// This structure stores the value of a given tag
typedef struct {
  BYTE tag;       // Tag to identify contents of item
  WORD length;    // Length of the item data
  BYTE *data;     // The block of data to decrypt or sign
} kssl_item;

// Possible values for KSSL_TAG_OPCODE

// A test message which will be echoed with its payload with the
// operation changed to OP_PONG
#define KSSL_OP_PING 0xF1
#define KSSL_OP_PONG 0xF2

// Decrypt data encrypted using RSA with or without RSA_PKCS1_PADDING
#define KSSL_OP_RSA_DECRYPT          0x01
#define KSSL_OP_RSA_DECRYPT_RAW      0x08

// Sign data using RSA
#define KSSL_OP_RSA_SIGN_MD5SHA1     0x02
#define KSSL_OP_RSA_SIGN_SHA1        0x03
#define KSSL_OP_RSA_SIGN_SHA224      0x04
#define KSSL_OP_RSA_SIGN_SHA256      0x05
#define KSSL_OP_RSA_SIGN_SHA384      0x06
#define KSSL_OP_RSA_SIGN_SHA512      0x07

// Sign data using ECDSA
#define KSSL_OP_ECDSA_MASK           0x10
#define KSSL_OP_ECDSA_SIGN_MD5SHA1   0x12
#define KSSL_OP_ECDSA_SIGN_SHA1      0x13
#define KSSL_OP_ECDSA_SIGN_SHA224    0x14
#define KSSL_OP_ECDSA_SIGN_SHA256    0x15
#define KSSL_OP_ECDSA_SIGN_SHA384    0x16
#define KSSL_OP_ECDSA_SIGN_SHA512    0x17

// Used to send a block of data back to the client (in response, for
// example, to a KSSL_OP_RSA_DECRYPT)
#define KSSL_OP_RESPONSE             0xF0
#define KSSL_OP_ERROR                0xFF

// Some error occurred, explanation is single byte in payload

typedef enum {
  KSSL_ERROR_NONE              = 0x00,
  KSSL_ERROR_CRYPTO_FAILED     = 0x01,
  KSSL_ERROR_KEY_NOT_FOUND     = 0x02,
  KSSL_ERROR_READ              = 0x03,
  KSSL_ERROR_VERSION_MISMATCH  = 0x04,
  KSSL_ERROR_BAD_OPCODE        = 0x05,
  KSSL_ERROR_UNEXPECTED_OPCODE = 0x06,
  KSSL_ERROR_FORMAT            = 0x07,
  KSSL_ERROR_INTERNAL          = 0x08
} kssl_error_code;

#endif // INCLUDED_KSSL

